ts/third_party/cloud-debug-nodejs/src/agent/io/scanner.ts

```diff
@@ -0,0 +1,264 @@
+  /**
+   * Encapsulates the results of a filesystem scan with methods
+   * to easily select scan information or filenames for a
+   * specific subset of the files listed in the scan results.
+   *
+   * @param stats An object that contains filenames
+   *  as keys where each key maps to an object containing the
+   *  hash and number of lines for the specified file.  This
+   *  information is accessed via the `hash` and `lines`
+   *  attributes respectively
+   * @param hash A hashcode computed from the contents of all the files.
+   */
+  constructor(
+      private readonly stats: ScanStats, readonly errorMap: Map<string, Error>,
+      readonly hash?: string) {}
+
+  errors(): Map<string, Error> {
+    return this.errorMap;
+  }
+
+  /**
+   * Used to get all of the file scan results.
+   */
+  all(): ScanStats {
+    return this.stats;
+  }
+
+  /**
+   * Used to get the file scan results for only the files
+   * whose filenames match the specified regex.
+   *
+   * @param {regex} regex The regex that tests a filename
+   *  to determine if the scan results for that filename
+   *  should be included in the returned results.
+   */
+  selectStats(regex: RegExp): ScanStats|{} {
+    return _.pickBy(this.stats, (ignore, key) => {
+      return regex.test(key);
+    });
+  }
+
+  /**
+   * Used to get the only the file paths in the scan results
+   * where the filenames match the specified regex and are
+   * returned with the each relative to the specified base
+   * directory.
+   *
+   * @param {regex} regex The regex that tests a filename to
+   *  determine if the scan results for that filename should
+   *  be included in the returned results.
+   * @param {string} baseDir The absolute path to the directory
+   *  from which all of the returned paths should be relative
+   *  to.
+   */
+  selectFiles(regex: RegExp, baseDir: string): string[] {
+    // ensure the base directory has only a single trailing path separator
+    baseDir = path.normalize(baseDir + path.sep);
+    return Object.keys(this.stats)
+        .filter((file) => {
+          return file && regex.test(file);
+        })
+        .map((file) => {
+          return path.normalize(file).replace(baseDir, '');
+        });
+  }
+}
+
+export async function scan(
+    shouldHash: boolean, baseDir: string, regex: RegExp): Promise<ScanResults> {
+  const fileList = await findFiles(baseDir, regex);
+  return await computeStats(fileList, shouldHash);
+}
+
+/**
+ * This function accept an array of filenames and computes a unique hash-code
+ * based on the contents.
+ *
+ * @param {!Array<string>} fileList array of filenames
+ * @param {Boolean} shouldHash whether a hash should be computed
+ * @param {!function(?Error, ?string, Object)} callback error-back style callback
+ *    returning the hash-code and an object containing file statistics.
+ */
+// TODO: Typescript: Fix the docs associated with this function to match the
+// call signature
+function computeStats(
```

Comment:
Can you just make this an async function, and drop the promise wrapping?

Reply:
This file is copied without modifications from https://github.com/googleapis/cloud-debug-nodejs.
cc/ @DominicKramer

Reply:
Even if copied, I still think we should do this correctly, no?

Reply:
I think this is best addressed in cloud-debug-nodejs. 
I have this file copied in as a subtree. To keep this maintainable, I really don't want to add changes to this.

Reply:
ahhhh, got it!